### PR TITLE
Crafting

### PR DESCRIPTION
**Pull request by overlordbleck**
image

### CODE REVIEW

File: gamemode/config/crafting.lua [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        name = "Furniture",
+        icon = "rvr/icons/craftingmenu_category_furniture.png",
+        recipes = {},
+    },
+    {
+        name = "Other",
+        icon = "rvr/icons/craftingmenu_category_other.png",
+        recipes = {},
+    },
+}
```
Comment (overlordbleck) [inline gamemode/config/crafting.lua]: WTF

File: gamemode/modules/crafting/sh_crafting.lua [hunk 2]
```diff
@@ -0,0 +1,31 @@
+RVR.Crafting = RVR.Crafting or {}
+local cft = RVR.Crafting
+
+cft.STATE_WAITING = 0
+cft.STATE_CRAFT_REQUEST = 1
+cft.STATE_CRAFTING = 2
+cft.STATE_CRAFTED = 3
+cft.STATE_GRAB_REQUEST = 4
```
Comment (overlordbleck) [inline gamemode/modules/crafting/sh_crafting.lua:8]: Why are these defined in here?

File: gamemode/modules/crafting/sh_crafting.lua [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+        if recipe.tier < categoryTier then
+            categoryTier = recipe.tier
+        end
+
+        recipe.categoryID = catID
+        recipe.recipeID = recipeID
+
+        recipe.itemsStruct = {}
+        for name, count in pairs( recipe.ingredients ) do
+            table.insert( recipe.itemsStruct, { item = { type = name }, count = count } )
```
Comment (overlordbleck) [inline gamemode/modules/crafting/sh_crafting.lua]: Smh, you don't have trailing commas here, how inconsistent.

File: gamemode/modules/crafting/sv_crafting.lua [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+util.AddNetworkString( "RVR_Crafting_AttemptCraft" )
+util.AddNetworkString( "RVR_Crafting_CraftResponse" )
+util.AddNetworkString( "RVR_Crafting_CloseCraftingMenu" )
+util.AddNetworkString( "RVR_Crafting_OpenCraftingMenu" )
+util.AddNetworkString( "RVR_Crafting_Grab" )
+
+local function craftFail( ply )
+    net.Start( "RVR_Crafting_CraftResponse" )
+    net.WriteInt( cft.STATE_WAITING, 4 )
```
Comment (overlordbleck) [inline gamemode/modules/crafting/sv_crafting.lua]: Didn't we agree to indent these?

File: gamemode/modules/crafting/sv_crafting.lua [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+            else
+                state = cft.STATE_CRAFTED
+            end
+        else
+            state = cft.STATE_CRAFTING
+        end
+    end
+
+    net.Start( "RVR_Crafting_CraftResponse" )
+    net.WriteInt( state, 4 )
```
Comment (overlordbleck) [inline gamemode/modules/crafting/sv_crafting.lua]: indent

File: gamemode/modules/crafting/sv_crafting.lua [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    local success = RVR.Inventory.tryTakeItems( ply, recipe.itemsStruct )
+    if not success then return end
+
+    ent.RVR_Crafting.output = {
+        recipe = recipe,
+        timeStart = CurTime(),
+    }
+
+    net.Start( "RVR_Crafting_CraftResponse" )
+    net.WriteInt( cft.STATE_CRAFTING, 4 )
```
Comment (overlordbleck) [inline gamemode/modules/crafting/sv_crafting.lua]: indent

File: gamemode/modules/crafting/sv_crafting.lua [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+        craftData.categoryID = recipe.categoryID
+        craftData.recipeID = recipe.recipeID
+    end
+
+    craftData.tier = ent.RVR_Crafting.tier
+    craftData.name = ent.RVR_Crafting.name
+    craftData.ent = ent
+
+    net.Start( "RVR_Crafting_OpenCraftingMenu" )
+    net.WriteTable( craftData )
```
Comment (overlordbleck) [inline gamemode/modules/crafting/sv_crafting.lua]: indent

**Review by overlordbleck — COMMENTED** [review]
Wow

File: gamemode/modules/crafting/sh_crafting.lua (hunk 3 above)
Comment (plally) [reply gamemode/modules/crafting/sh_crafting.lua]: you only use trailing commas on multi line tables, when theres a newline after the comma.

File: gamemode/modules/crafting/sh_crafting.lua (hunk 2 above)
Comment (overlordbleck) [reply gamemode/modules/crafting/sh_crafting.lua:8]: w/e it's fine here

Comment (samuelWilliams99) [reply gamemode/modules/crafting/sh_crafting.lua:8]: Idk where else i could put them

File: gamemode/config/crafting.lua [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+        ingredient1Type = count1
+        ingredient12ype = count2
+    }
+    tier = (optional) Minimum crafter tier required, defaults to 1
+    timeToCraft = Time in seconds this craft takes (non-zero)
+}
+
+]]
+
+RVR.Crafting.Recipes = {
```
Comment (plally) [inline gamemode/config/crafting.lua]: lets have functions to add stuff  to this

File: gamemode/modules/crafting/sh_crafting.lua [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+local cft = RVR.Crafting
+
+cft.STATE_WAITING = 0
+cft.STATE_CRAFT_REQUEST = 1
+cft.STATE_CRAFTING = 2
+cft.STATE_CRAFTED = 3
+cft.STATE_GRAB_REQUEST = 4
+
+for catID, category in ipairs( cft.Recipes ) do
+    local categoryTier = 1000
```
Comment (plally) [inline gamemode/modules/crafting/sh_crafting.lua]: this name is a bit confusing

File: gamemode/modules/crafting/sv_crafting.lua [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+    ent.RVR_Crafting.output = {
+        recipe = recipe,
+        timeStart = CurTime(),
+    }
+
+    net.Start( "RVR_Crafting_CraftResponse" )
+        net.WriteInt( cft.STATE_CRAFTING, 4 )
+    net.Send( ply )
+
+    return true
```
Comment (plally) [inline gamemode/modules/crafting/sv_crafting.lua:83]: if this returns  a boolean, the other returns should return false instead of nil

**Review by plally — CHANGES_REQUESTED** [review]

**Review by plally — APPROVED** [review]

**Review by ghost — APPROVED** [review]
Looks good to me

**Pull request merged**